storage/innobase/xtrabackup/test/t/dictionary.sh

```diff
@@ -0,0 +1,291 @@
+. inc/common.sh
+
+vlog "#"
+vlog "# PXB-2955 : Implement dictionary cache"
+vlog "#"
+
+function wait_for_bg_trx() {
+
+ while true; do
+   mysql -e 'SHOW PROCESSLIST' |  grep -i 'SLEEP'
+   if [ $? -eq 0 ]; then
+	   break;
+   fi
+   vlog "waiting for background trx to finish inserting data and reach to sleep"
+   sleep 1
+ done
+
+ vlog "Wait for trx data to be flushed. This guarantees rollback during prepare"
+ innodb_wait_for_flush_all
+}
+
+function kill_bg_trx() {
+ TRX_PID=$1
+ vlog "Kill the Long running uncommited transaction after the backup"
+ kill_query_pattern "INFO like '%SLEEP%'"
+
+ kill -SIGKILL $TRX_PID
+
+ vlog "Wait for killed process to exit"
+ wait $TRX_PID
+}
+
+function check_rollback_msg() {
+FILE=$1
+egrep -q "Rolling back trx with id [0-9]+, [0-9]+ rows to undo" $FILE || die "Expected rollback didn't happen during prepare"
+}
+
+function create_tablespace() {
+ mysql test << EOF
+ CREATE TABLESPACE ts1 ADD DATAFILE 'ts1.ibd';
+ CREATE TABLE t11(a INT) TABLESPACE=ts1;
+ CREATE TABLE t12(a INT, b INT) TABLESPACE=ts1;
+EOF
+}
+
+function create_tables() {
+  TABLE=$1
+  # optional parameter 2 for TABLESPACE
+  TABLESPACE=${2:-}
+  if [ ! -z ${TABLESPACE} ];
+  then
+    TABLESPACE="TABLESPACE=$TABLESPACE"
+  fi
+  mysql test << EOF
+    CREATE TABLE $TABLE (
+    id INT PRIMARY KEY,
+    val INT NOT NULL DEFAULT 0
+    ) $TABLESPACE ENGINE=InnoDB
+EOF
+
+  mysql -e "INSERT INTO $TABLE (id) VALUES (1),(2),(3),(4),(5),(6),(7),(8),(9),(10)" test
+  vlog "Long running transaction that modifies table data but is not committed at the time of backup"
+  $MYSQL $MYSQL_ARGS -e "\
+   BEGIN;\
+   INSERT INTO $TABLE (id) SELECT id + (SELECT MAX(id) FROM $TABLE) FROM $TABLE;\
+   INSERT INTO $TABLE (id) SELECT id + (SELECT MAX(id) FROM $TABLE) FROM $TABLE;\
+   INSERT INTO $TABLE (id) SELECT id + (SELECT MAX(id) FROM $TABLE) FROM $TABLE;\
+   INSERT INTO $TABLE (id) SELECT id + (SELECT MAX(id) FROM $TABLE) FROM $TABLE;\
+   INSERT INTO $TABLE (id) SELECT id + (SELECT MAX(id) FROM $TABLE) FROM $TABLE;\
+   INSERT INTO $TABLE (id) SELECT id + (SELECT MAX(id) FROM $TABLE) FROM $TABLE;\
+   INSERT INTO $TABLE (id) SELECT id + (SELECT MAX(id) FROM $TABLE) FROM $TABLE;\
+   INSERT INTO $TABLE (id) SELECT id + (SELECT MAX(id) FROM $TABLE) FROM $TABLE;\
+   INSERT INTO $TABLE (id) SELECT id + (SELECT MAX(id) FROM $TABLE) FROM $TABLE;\
+   INSERT INTO $TABLE (id) SELECT id + (SELECT MAX(id) FROM $TABLE) FROM $TABLE;\
+   INSERT INTO $TABLE (id) SELECT id + (SELECT MAX(id) FROM $TABLE) FROM $TABLE;\
+   SELECT SLEEP (10000);
+ " test &
+ uncommitted_trx=$!
+
+ wait_for_bg_trx
+
+ vlog "Backup"
+ rm -rf $topdir/backup
+ xtrabackup --backup --target-dir=$topdir/backup 2>&1 | tee $topdir/pxb.log
```

Comment:
Can we add a trap to execute `kill_bg_trx $uncommitted_trx` in case xtrabackup fails, we will never reach `kill_bg_trx $uncommitted_trx`

Reply:
Done